### PR TITLE
test(loop): cover exit-condition termination (issue #187)

### LINKED ISSUE #187: test(core-components): extend loop-component-regression with exit-condition test
## Context

`core-components/` Core Components — section **3.6 Loop Component** in `QA-CHECKLIST.md` has 1 `[ ]` left:

> Loop stops when exit condition is met

The existing `loop-component-regression.spec.ts` already covers handles, output buttons, build-failure, and the Research Translation Loop template iteration. This issue tracks adding a single test that explicitly validates the loop's exit-condition logic (independent of the template iteration count).

## Scope (single test)

Build a minimal loop with a configurable exit condition (or use a template), run it, and assert iteration stops as soon as the condition is met.

Approach options to evaluate during implementation:
- Use a custom-component condition that flips after N iterations
- Or rely on a starter template that exposes an explicit exit condition

## Tags

`@stable @regression @components`.

## Deliverables

- [ ] New test added to `tests/tests-automations/regression/core-components/loop-component-regression.spec.ts`
- [ ] `docs/core-components/loop-component-regression.md` updated to describe the new test
- [ ] `QA-CHECKLIST.md` — section 3.6 bullet switched from `[ ]` to `[x]`

### PR DESCRIPTION
**Pull request by daniellicnerski1**
## Type

- [x] `feat` — new test for the Loop component's exit-condition behavior

---

## What this PR does

Closes the last `[ ]` of `QA-CHECKLIST.md` section 3.6 (Loop Component) by adding a deterministic exit-condition test:

- New asset `tests/assets/flows/loop-exit-condition.json` — pre-built 4-node flow (Create List + Loop + Type Convert + Chat Output) wired in a loop cycle. Default `texts` is a placeholder; the test mutates it per scenario.
- New test in `core-components/loop-component-regression.spec.ts`: `Loop component — stops after exhausting input DataFrame and emits aggregated done`. Single `test()` with two `test.step` blocks (N=3 and N=1). Tags `@stable @regression @components`.

The Loop component has no explicit "exit condition" input — termination is implicit (loop finishes when the input DataFrame is exhausted). The test asserts `done.length === input.length` for two N values, proving termination behavior is correct.

---

## Tests covered

| # | Test | What it validates |
|---|---|---|
| 1 | `Loop component — stops after exhausting input DataFrame and emits aggregated done` (`N=3`) | Loop iterates exactly 3 times when input DataFrame has 3 rows; catches off-by-one (stopping at N-1 or running N+1) |
| 2 | (same `test()`, second `test.step` `N=1`) | Smallest non-empty case; catches bugs where body is skipped for single-item DataFrames or `done` fires before the last iteration |

---

## How the tests were built

The flow is created server-side via `POST /api/v1/flows/` with the asset JSON as the body, with `texts` patched in-memory to control N. Avoids two failure modes observed during implementation:

1. **Drag-drop upload race** — `simulateDragAndDrop` produced a flow that the home page card list would not reliably refresh in time (~50% flake). API creation is deterministic.
2. **Canvas drag-and-drop fragility** — building the flow in-test by adding components from the sidebar would have required wiring handles via `browser_drag`, which fails when sibling nodes overlap (Langflow places new components at the same default position).

After `POST 201`, navigation uses **click-on-card** rather than `goto(/flow/{id})`. Deep-linking intermittently redirected to the flows list when the React app's owned-flows cache was stale at navigation time (~30% flake observed). Navigating to `/` first triggers a full owned-flows re-fetch, so the click resolves to the canvas deterministically.

Done-output assertion uses the paginated treegrid pagination summary (`"1 to N of N. Page 1 of 1"`) parsed via regex — robust to row sorting and avoids brittle DOM-row counting under virtualization.

N=0 is intentionally out of scope: source code (`loop.py:181-184`) shows that empty input takes a separate code path (initialization short-circuit, `self.log("No items to iterate, skipping loop body")`), distinct from termination by exhaustion.

---

## Dependencies

- Asset file `tests/assets/flows/loop-exit-condition.json` (committed in this PR).
- No `OPENAI_API_KEY` or other LLM provider required — the flow is fully deterministic.
- Loop component must expose `inputs`, `item`, `done` ports per current source.
- Create List component must offer the `dataframe` (Table) output via the dropdown — confirmed in Langflow 1.10.x.
- Test runs in `serial` mode (already configured at file level for the existing Loop tests).

---

## What this PR does not cover

- N=0 edge case (separate code path: Loop initialization short-circuit when input is empty; not termination semantics).
- Conditional early-exit triggered from inside the body — Langflow's Loop has no such mechanism.
- The 3 existing Loop tests are not modified.

---

## Related issue

Closes #187.

---

## Validation

- [x] `npx tsc --noEmit` — clean
- [x] `npx eslint` on the file — 0 errors (13 pre-existing warnings only)
- [x] `playwright-test-linter` skill — all anti-pattern checks pass
- [x] 10× consecutive `--retries=0` runs — **10/10 pass** (~10s avg)
- [x] Force-fail (`toBe(999)` instead of `toBe(3)`): assertion fails as expected, then reverted
- [x] `--trace=on` full spec run — 4/4 passed (58.6s)
- [x] 0 occurrences of `🚨 Backend Error` in `test-results/`
- [x] `QA-CHECKLIST.md` section 3.6 bullet flipped to `[x]`

### CODE REVIEW

File: tests/tests-automations/regression/core-components/loop-component-regression.spec.ts [hunk 1]
```diff
@@ -202,3 +204,136 @@ test(
...
+      if (token) xhr.setRequestHeader("Authorization", `Bearer ${token}`);
+      xhr.onload = () => {
+        if (xhr.status !== 201)
+          return reject(`status ${xhr.status}: ${xhr.responseText.slice(0, 200)}`);
+        resolve();
+      };
+      xhr.onerror = () => reject("network error");
+      xhr.send(JSON.stringify(flowBody));
+    });
+  }, flow);
```
[the remark this answers is not in the document]
Comment (daniellicnerski1) [reply tests/tests-automations/regression/core-components/loop-component-regression.spec.ts]: Addressed in commit aee373d. `createFlowFromAsset` now uses `request.post('/api/v1/flows/', { headers, data })` with the existing `getAuthToken` helper — same pattern as `cleanAllFlows` and the `api/flows` specs. Removed the in-browser `page.evaluate` + `XMLHttpRequest` plumbing entirely. The original concern about Langflow's bundled HTTP wrapper / Accept-Language `TypeError` was specific to the page's `fetch` — Playwright's request context is a separate test-runner HTTP client and not affected. The function now returns the created flow id so callers can target cleanup or deep-link.

File: tests/tests-automations/regression/core-components/loop-component-regression.spec.ts [hunk 2]
```diff
@@ -202,3 +204,136 @@ test(
...
+      xhr.open("POST", "/api/v1/flows/", true);
+      xhr.withCredentials = true;
+      xhr.setRequestHeader("Content-Type", "application/json");
+      if (token) xhr.setRequestHeader("Authorization", `Bearer ${token}`);
+      xhr.onload = () => {
+        if (xhr.status !== 201)
+          return reject(`status ${xhr.status}: ${xhr.responseText.slice(0, 200)}`);
+        resolve();
+      };
+      xhr.onerror = () => reject("network error");
```
[the remark this answers is not in the document]
Comment (daniellicnerski1) [reply tests/tests-automations/regression/core-components/loop-component-regression.spec.ts]: Addressed in commit aee373d. The `page.evaluate` + `XMLHttpRequest` block was removed entirely as part of the switch to `request.post()`. On a non-201 response the function now throws an `Error` with the HTTP status plus a truncated response body (`POST /api/v1/flows/ failed: status ${status}: ${body}`), so Playwright keeps a real stack/context.

File: docs/core-components/loop-component-regression.md [hunk 3]
```diff
@@ -6,11 +6,12 @@
 
 ## What this test validates *(required)*
 
-Validates three fundamental behaviors of the Loop component on the Langflow canvas:
+Validates four fundamental behaviors of the Loop component on the Langflow canvas:
 
 1. **Correct rendering** — the node appears on the canvas with all expected handles (`inputs-left`, `item-left`, `item-right`, `done-right`) and with the output inspection buttons present in the node footer.
 2. **Error path without connections** — running the Loop with no connections shows the "Flow build failed" notification without freezing the interface; the node remains intact and the run button stays accessible.
 3. **Real iteration via template** — using the "Research Translation Loop" template, the Loop iterates over 2 ArXiv articles and produces an aggregated response in the Playground containing at least 2 mentions of "Title", confirming the loop completed both iterations.
```
[the remark this answers is not in the document]
Comment (daniellicnerski1) [reply docs/core-components/loop-component-regression.md]: Addressed in commit aee373d. Bullet 3 of 'What this test validates' on `docs/core-components/loop-component-regression.md` now states "at least 1 mention of 'Title'" to match the spec assertion (`titleCount >= 1`) and explains why the threshold is intentionally relaxed (LLM non-determinism) — with a pointer that iteration count itself is covered deterministically by Test 4.

**Comment by daniellicnerski1** [comment]
## Copilot review addressed — commit aee373d

All 3 comments from Copilot's review have been addressed in a single follow-up commit. Summary:

### 1. `createFlowFromAsset` rewritten to use `page.request` + `getAuthToken` (Copilot comment #1)

The function no longer runs an in-browser `page.evaluate` + `XMLHttpRequest` block. It now uses Playwright's request context, mirroring the pattern already used by `cleanAllFlows` and the `api/flows` specs:

```ts
async function createFlowFromAsset(
  request: APIRequestContext,
  flow: Record<string, unknown>,
): Promise<string> {
  const authToken = await getAuthToken(request);
  const res = await request.post("/api/v1/flows/", {
    headers: authToken ? { Authorization: authToken } : {},
    data: flow,
  });
  if (res.status() !== 201) {
    const body = (await res.text()).slice(0, 200);
    throw new Error(`POST /api/v1/flows/ failed: status ${res.status()}: ${body}`);
  }
  const body = (await res.json()) as { id: string };
  return body.id;
}
```

The function now also returns the created flow id, so callers can target cleanup or deep-link if needed. The original Accept-Language `TypeError` concern from the XHR-era comment was specific to the page's `fetch` — Playwright's request context is a separate test-runner HTTP client and not affected.

### 2. Error rejection now uses `Error` instances (Copilot comment #2)

The `page.evaluate` block was removed entirely as part of the switch above, so the string-`reject` issue is gone. Non-201 responses now `throw new Error(...)` with the HTTP status plus a truncated response body, so Playwright keeps a real stack/context for debugging.

### 3. Doc bullet 3 in `docs/core-components/loop-component-regression.md` (Copilot comment #3)

"At least 2 mentions of 'Title'" → "at least 1 mention of 'Title'", matching the spec's `titleCount >= 1` assertion. Added a sentence explaining the relaxed threshold (LLM non-determinism) and a pointer that iteration count itself is now covered deterministically by Test 4.

### Validation

- `npm run typecheck` — clean
- `npx eslint` on the file — 0 errors (13 pre-existing warnings only, unchanged)
- CI: TypeScript Check + ESLint both green on the new commit
